test/aave-v3/TestSupplyVault.t.sol

```diff
@@ -490,27 +490,100 @@ contract TestSupplyVault is TestSetupVaults {
+            address(daiSupplyVault),
+            rewardToken
+        );
+        morpho.updateIndexes(aDai);
+
+        // Should update the unclaimed amount
+        vaultSupplier2.mintVault(
+            daiSupplyVault,
+            daiSupplyVault.previewMint(amount),
+            address(vaultSupplier1)
+        );
+        (, uint128 userReward) = daiSupplyVault.userRewards(rewardToken, address(vaultSupplier1));
+        assertApproxEqAbs(uint256(userReward), expectedTotalRewardsAmount, 10000);
+    }
+
+    function testRewardsShouldAccrueWhenRedeemingOnBehalf() public {
+        uint256 amount = 10_000 ether;
+        address[] memory poolTokens = new address[](1);
+        poolTokens[0] = aDai;
+
+        vaultSupplier1.depositVault(daiSupplyVault, amount);
+        vm.roll(block.number + 100);
+
+        uint256 expectedTotalRewardsAmount = rewardsManager.getUserRewards(
+            poolTokens,
+            address(daiSupplyVault),
+            rewardToken
+        );
+        morpho.updateIndexes(aDai);
 
-        daiSupplyVault.redeem(
-            daiSupplyVault.balanceOf(address(supplier1)),
-            address(supplier2),
-            address(supplier1)
+        // Should update the unclaimed amount
+        vaultSupplier1.redeemVault(
+            daiSupplyVault,
+            daiSupplyVault.balanceOf(address(vaultSupplier1)),
+            address(vaultSupplier2),
+            address(vaultSupplier1)
         );
-        vm.stopPrank();
+        (, uint128 userReward) = daiSupplyVault.userRewards(rewardToken, address(vaultSupplier1));
+        assertApproxEqAbs(uint256(userReward), expectedTotalRewardsAmount, 10000);
+    }
 
-        // Balance of supplier1 is expected to be 0.
-        assertEq(daiSupplyVault.balanceOf(address(supplier1)), 0);
+    function testRewardsShouldAccrueWhenWithdrawingOnBehalf() public {
+        uint256 amount = 10_000 ether;
+        address[] memory poolTokens = new address[](1);
+        poolTokens[0] = aDai;
+
+        vaultSupplier1.depositVault(daiSupplyVault, amount);
+        vm.roll(block.number + 100);
 
-        // supplier1 must have some rewards to claim.
-        (, uint256[] memory unclaimed) = daiSupplyVault.getAllUnclaimedRewards(address(supplier1));
-        assertGt(unclaimed[0], 0);
+        uint256 expectedTotalRewardsAmount = rewardsManager.getUserRewards(
+            poolTokens,
+            address(daiSupplyVault),
+            rewardToken
+        );
+        morpho.updateIndexes(aDai);
+
+        // Should update the unclaimed amount
+        vaultSupplier1.withdrawVault(
+            daiSupplyVault,
+            daiSupplyVault.previewWithdraw(daiSupplyVault.balanceOf(address(vaultSupplier1))),
+            address(vaultSupplier2),
+            address(vaultSupplier1)
+        );
+        (, uint128 userReward) = daiSupplyVault.userRewards(rewardToken, address(vaultSupplier1));
+        assertApproxEqAbs(uint256(userReward), expectedTotalRewardsAmount, 10000);
```

Comment:
I think we should check that vaultSupplier2's index has not changed. Same for redeem since this was our issue.